lib/evm/interpreter.js

```diff
@@ -0,0 +1,206 @@
+const promisify = require('util.promisify')
+const BN = require('bn.js')
+const ethUtil = require('ethereumjs-util')
+const { ERROR } = require('../exceptions')
+const Loop = require('./loop')
+const { StorageReader } = require('../state')
+const PStateManager = require('../state/promisified')
+
+module.exports = class Interpreter {
+  constructor (vm, txContext, block, storageReader) {
+    this._vm = vm
+    this._state = new PStateManager(this._vm.stateManager)
+    this._storageReader = storageReader || new StorageReader(this._state._wrapped)
+    this._tx = txContext
+    this._block = block
+  }
+
+  async executeMessage (message) {
+    await this._state.checkpoint()
+
+    const account = await this._state.getAccount(message.caller)
+    // Reduce tx value from sender
+    await this._reduceSenderBalance(account, message)
+
+    // Load `to` account
+    let toAccount
+    let createdAddress
+    if (message.to) {
+      toAccount = await this._state.getAccount(message.to)
+    } else {
+      ({ toAccount, createdAddress } = await this._createContract(account, message, message.salt))
+    }
+
+    // Add tx value to the `to` account
+    await this._addToBalance(toAccount, message)
+
+    // Load code
+    await this._loadCode(message)
+    if (!message.code || message.code.length === 0) {
+      await this._state.commit()
+      return {
+        gasUsed: new BN(0),
+        createdAddress: createdAddress,
+        vm: {
+          exception: 1
+        }
+      }
+    }
+
+    let { err, results } = await this.runLoop(message)
+    results = await this._parseRunResult(err, results, message, createdAddress)
+
+    // Save code if a new contract was created
+    if (createdAddress && !results.runState.vmError && results.return && results.return.toString() !== '') {
+      await this._state.putContractCode(createdAddress, results.return)
+    }
+
+    return {
+      gasUsed: results.gasUsed,
+      createdAddress: createdAddress,
+      vm: results
+    }
+  }
+
+  async runLoop (message) {
+    const opts = {
+      storageReader: this._storageReader,
+      block: this._block,
+      gasPrice: this._tx.gasPrice,
+      origin: this._tx.origin,
+      code: message.code,
+      data: message.data,
+      gasLimit: message.gasLimit,
+      address: message.to,
+      caller: message.caller,
+      value: message.value,
+      depth: message.depth,
+      selfdestruct: message.selfdestruct,
+      static: message.isStatic
+    }
```

Comment:
What is the reason for this split of txContext and message? Couldn't this be combined in the first place, since this is combined here again?

Reply:
Splitting this `opts` into `message` and `txContext` (+ `storageReader` and `block` for now) is something I'm planning to do along with splitting `runState`.

Reply:
Ah ok.